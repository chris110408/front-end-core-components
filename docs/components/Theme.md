# Theme

The Theme component provides global theme configuration state via React's [context API](https://reactjs.org/docs/context.html). It has support built in for:

* Multiple themes
* Persisting to local storage

Each of the above features are optional.

## Dependencies

* react `^16.3`
* prop-types `^15.6.2`

## Reference

| prop name | description | type | required? | default value |
| --------- | ----------- | ---- | --------- | ------------- |
| themes | Array of themes to choose from | `Array<Object>` | ✓ | - |
| themes.key | Reference for a theme | `String` | ✓ | - |
| themes.config | The themes configuration, any structure for your use case can be accepted | `Object` | ✓ | - |
| onUpdate | Called whenever the theme is updated with the following arguments - theme, prevTheme. prevTheme will be null when the component mounts. | ✗ | - |
| persistTheme | Whether to persist the selected theme to local storage or not and also retrieve values from local storage | `Boolean` | ✗ | true |
| transformTheme | Called before the theme is output to the consumer, defaults to returning `theme.config` accepts single argument of `theme` | `Function` | ✗ | (theme) => theme.config |

## Usage

### Provider

```javascript
import React from 'react'
import Theme from '@input-output-hk/front-end-core-components/components/Theme'

const App = () => (
  <Location>
    {({ location }) => (
      <Theme.Provider
        themes={[
          {
            key: 'iohk-dark',
            config: {
              name: 'Dark',
              colors: {
                primary: 'tomato',
                secondary: 'papayawhip'
              },
              typography: {
                fontFamily: 'Arial'
              }
            }
          },
          {
            key: 'iohk-light',
            config: {
              name: 'Light',
              colors: {
                primary: 'thistle',
                secondary: 'dodgerblue'
              },
              typography: {
                fontFamily: 'Helvetica'
              }
            }
          }
        ]}
        onUpdate={({ theme, prevTheme }) => {
          // Optional callback whenever the theme is updated
          // When the component mounts `prevTheme` will be null
          // An alternative to using this prop would be to use the consumer
          // and useEffect(() => { ... }, [ key ]) where appropriate

          // Examples
          // * track event to analytics
          // * display an on screen confirmation
        }}
        persistTheme={
          true // defaults to true, whether to persist theme to local storage or not
        }
        transformTheme={({ key, config }) => {
          // Transforms the theme output to Consumer -> theme
          return config
        }}
      >
        <MyComponent />
      </Theme.Provider>
    )}
  </Location>
)

```

### Consumer

```javascript
import React from 'react'
import Theme from '@input-output-hk/front-end-core-components/components/Theme'

const MyComponent = () => (
  <Theme.Consumer>
    {({
      key,
      theme,
      // Un-transformed theme
      originalTheme,
      setTheme,
      themes
    }) => (
      <div>
        <select value={key} onChange={e => setTheme(e.target.value)}>
          {themes.map(({ key, config }) => (
            <option key={key} value={key}>
              {config.name}
            </option>
          ))}
        </select>
        <p>Selected theme, {theme.name}</p>
      </div>
    )}
  </Theme.Consumer>
)

```
