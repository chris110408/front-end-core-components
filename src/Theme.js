import React, { createContext, useState, useEffect } from 'react'
import PropTypes from 'prop-types'

const ThemeContext = createContext()
const Consumer = ThemeContext.Consumer

const Provider = ({
  children,
  themes,
  onUpdate = ({ theme, prevTheme }) => {},
  persistTheme = true
}) => {
  const [ theme, updateTheme ] = useState(getInitialTheme())

  useEffect(() => {
    persistToLocalStorage(theme)
    onUpdate({ theme, prevTheme: null })
  }, [])

  /**
   * Get the initial theme by first searching localStorage when
   * persistTheme is true, falling back to the first theme in the
   * themes array
   *
   * @returns {String} Theme key
   */
  function getInitialTheme () {
    let theme
    if (persistTheme && window && window.localStorage && window.localStorage.getItem) theme = window.localStorage.getItem('theme') || ''
    return isValidTheme(theme)
      ? theme
      : themes[0].key
  }

  /**
   * Persists the theme to local storage if persistTheme is true
   *
   * @param {String} theme
   * @returns {Void}
   */
  function persistToLocalStorage (theme) {
    if (persistTheme && window && window.localStorage && window.localStorage.setItem) window.localStorage.setItem('theme', theme)
  }

  /**
   * Checks if a theme exists i.e. is valid
   *
   * @param {String} theme Theme key
   * @returns {Boolean}
   */
  function isValidTheme (theme) {
    return Boolean(themes.filter(({ key }) => key === theme).shift())
  }

  /**
   * Sets a new theme, will do nothing when setting the same theme
   *
   * @param {String} newTheme Theme key
   * @returns {Void}
   */
  function setTheme (newTheme) {
    if (newTheme === theme) return
    if (!isValidTheme(newTheme)) throw new Error(`Theme ${newTheme} does not exist`)
    persistToLocalStorage(newTheme)
    onUpdate({ theme: newTheme, prevTheme: theme })
    updateTheme(theme)
  }

  /**
   * Gets the theme configuration for the theme key
   *
   * @returns {Object}
   */
  function getTheme () {
    return (themes.filter(({ key }) => key === theme).shift() || {})
  }

  return (
    <ThemeContext.Provider
      value={{
        theme: getTheme(),
        setTheme,
        themes
      }}
    >
      {children}
    </ThemeContext.Provider>
  )
}

Provider.propTypes = {
  children: PropTypes.any,
  themes: PropTypes.arrayOf(PropTypes.shape({
    key: PropTypes.string.isRequired,
    config: PropTypes.object.isRequired
  })).isRequired,
  onUpdate: PropTypes.func,
  persistTheme: PropTypes.bool
}

export {
  Provider,
  Consumer
}
