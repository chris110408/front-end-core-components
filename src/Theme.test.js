import React from 'react'
import { Provider, Consumer } from './Theme'
import { mount, shallow } from 'enzyme'
import { act } from 'react-dom/test-utils'

const DEFAULT_PROVIDER_PROPS = {
  themes: [
    {
      key: 'iohk-dark',
      config: {
        colors: {
          primary: 'tomato',
          secondary: 'papayawhip'
        },
        typography: {
          fontFamily: 'Arial'
        }
      }
    },
    {
      key: 'iohk-light',
      config: {
        colors: {
          primary: 'thistle',
          secondary: 'dodgerblue'
        },
        typography: {
          fontFamily: 'Helvetica'
        }
      }
    }
  ],
  children: (<p>Theme provider children</p>)
}

function getProviderWrapper ({ props = {}, renderFunction }) {
  return renderFunction(
    <Provider
      themes={props.themes || DEFAULT_PROVIDER_PROPS.themes}
      onUpdate={props.onUpdate}
      persistTheme={props.persistTheme}
    >
      {props.children ? props.children : DEFAULT_PROVIDER_PROPS.children}
    </Provider>
  )
}

describe('Theme context', () => {
  describe('<Provider />', () => {
    let provider
    beforeEach(() => {
      provider = getProviderWrapper({ renderFunction: shallow })
    })

    test('it renders the provider component correctly', () => {
      expect(provider).toMatchSnapshot()
    })

    describe('when the provider is mounted', () => {
      let onUpdate
      beforeEach(() => {
        onUpdate = jest.fn()
        provider = getProviderWrapper({ renderFunction: mount, props: { onUpdate } })
      })

      test('the theme is persisted to local storage', () => {
        expect(window.localStorage.setItem).toBeCalledWith('theme', 'iohk-dark')
      })

      test('onUpdate is called with the first theme in the themes list', () => {
        expect(onUpdate).toBeCalledWith({ theme: 'iohk-dark', prevTheme: null })
      })

      describe('when a theme pre-exists in local storage', () => {
        beforeEach(() => {
          onUpdate.mockReset()
          window.localStorage.getItem.mockImplementation((key) => key === 'theme' ? 'iohk-light' : undefined)
          provider = getProviderWrapper({ renderFunction: mount, props: { onUpdate } })
        })

        test('onUpdate is called with the theme from local storage', () => {
          expect(onUpdate).toBeCalledWith({ theme: 'iohk-light', prevTheme: null })
        })

        describe('when the theme in local storage is invalid', () => {
          beforeEach(() => {
            onUpdate.mockReset()
            window.localStorage.getItem.mockImplementation((key) => key === 'theme' ? 'iohk-orange' : undefined)
            provider = getProviderWrapper({ renderFunction: mount, props: { onUpdate } })
          })

          test('onUpdate is called with the first theme in the themes list', () => {
            expect(onUpdate).toBeCalledWith({ theme: 'iohk-dark', prevTheme: null })
          })
        })

        describe('when persistTheme is false', () => {
          beforeEach(() => {
            onUpdate.mockReset()
            window.localStorage.setItem.mockReset()
            window.localStorage.getItem.mockImplementation((key) => key === 'theme' ? 'iohk-light' : undefined)
            provider = getProviderWrapper({ renderFunction: mount, props: { onUpdate, persistTheme: false } })
          })

          test('onUpdate is called with the first theme in the themes list', () => {
            expect(onUpdate).toBeCalledWith({ theme: 'iohk-dark', prevTheme: null })
          })

          test('the theme is not persisted to local storage', () => {
            expect(window.localStorage.setItem).not.toBeCalled()
          })
        })
      })
    })
  })

  describe('<Consumer />', () => {
    let onUpdate, wrapper
    beforeEach(() => {
      onUpdate = jest.fn()
      wrapper = getProviderWrapper({
        renderFunction: mount,
        props: {
          onUpdate,
          children: (
            <Consumer>
              {({ key, theme, setTheme, themes }) => (
                <div
                  id='consumer-value-holder'
                  consumervalue={{
                    key,
                    theme,
                    setTheme,
                    themes
                  }}
                />
              )}
            </Consumer>
          )
        }
      })
    })

    afterEach(() => {
      onUpdate.mockReset()
    })

    test('the consumer values are correct', () => {
      expect(wrapper.find('#consumer-value-holder').prop('consumervalue')).toMatchSnapshot()
    })

    describe('when setTheme is called', () => {
      beforeEach(() => {
        act(() => {
          wrapper.find('#consumer-value-holder').prop('consumervalue').setTheme('iohk-light')
        })
        wrapper.update()
      })

      test('the new theme is persisted to local storage', () => {
        expect(window.localStorage.setItem).toBeCalledWith('theme', 'iohk-light')
      })

      test('the consumer values are correctly updated', () => {
        expect(wrapper.find('#consumer-value-holder').prop('consumervalue')).toMatchSnapshot()
      })

      test('when attempting to set an invalid theme', () => {
        expect(() => {
          wrapper.find('#consumer-value-holder').prop('consumervalue').setTheme('cardano')
        }).toThrow('Theme cardano does not exist')
      })
    })
  })
})
