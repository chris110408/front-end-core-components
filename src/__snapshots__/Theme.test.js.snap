// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme context <Consumer /> the consumer values are correct 1`] = `
Object {
  "key": "iohk-dark",
  "setTheme": [Function],
  "theme": Object {
    "colors": Object {
      "primary": "tomato",
      "secondary": "papayawhip",
    },
    "typography": Object {
      "fontFamily": "Arial",
    },
  },
  "themes": Array [
    Object {
      "config": Object {
        "colors": Object {
          "primary": "tomato",
          "secondary": "papayawhip",
        },
        "typography": Object {
          "fontFamily": "Arial",
        },
      },
      "key": "iohk-dark",
    },
    Object {
      "config": Object {
        "colors": Object {
          "primary": "thistle",
          "secondary": "dodgerblue",
        },
        "typography": Object {
          "fontFamily": "Helvetica",
        },
      },
      "key": "iohk-light",
    },
  ],
}
`;

exports[`Theme context <Consumer /> when setTheme is called the consumer values are correctly updated 1`] = `
Object {
  "key": "iohk-dark",
  "setTheme": [Function],
  "theme": Object {
    "colors": Object {
      "primary": "tomato",
      "secondary": "papayawhip",
    },
    "typography": Object {
      "fontFamily": "Arial",
    },
  },
  "themes": Array [
    Object {
      "config": Object {
        "colors": Object {
          "primary": "tomato",
          "secondary": "papayawhip",
        },
        "typography": Object {
          "fontFamily": "Arial",
        },
      },
      "key": "iohk-dark",
    },
    Object {
      "config": Object {
        "colors": Object {
          "primary": "thistle",
          "secondary": "dodgerblue",
        },
        "typography": Object {
          "fontFamily": "Helvetica",
        },
      },
      "key": "iohk-light",
    },
  ],
}
`;

exports[`Theme context <Consumer /> when the theme is transformed it renders with the transformed theme 1`] = `
<Provider
  themes={
    Array [
      Object {
        "config": Object {
          "colors": Object {
            "primary": "tomato",
            "secondary": "papayawhip",
          },
          "typography": Object {
            "fontFamily": "Arial",
          },
        },
        "key": "iohk-dark",
      },
      Object {
        "config": Object {
          "colors": Object {
            "primary": "thistle",
            "secondary": "dodgerblue",
          },
          "typography": Object {
            "fontFamily": "Helvetica",
          },
        },
        "key": "iohk-light",
      },
    ]
  }
  transformTheme={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "config": Object {
              "colors": Object {
                "primary": "tomato",
                "secondary": "papayawhip",
              },
              "typography": Object {
                "fontFamily": "Arial",
              },
            },
            "key": "iohk-dark",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Object {
            "colors": Object {
              "primary": "black",
              "secondary": "papayawhip",
            },
            "typography": Object {
              "fontFamily": "Arial",
            },
          },
        },
      ],
    }
  }
>
  <div
    consumervalue={
      Object {
        "key": "iohk-dark",
        "setTheme": [Function],
        "theme": Object {
          "colors": Object {
            "primary": "black",
            "secondary": "papayawhip",
          },
          "typography": Object {
            "fontFamily": "Arial",
          },
        },
        "themes": Array [
          Object {
            "config": Object {
              "colors": Object {
                "primary": "tomato",
                "secondary": "papayawhip",
              },
              "typography": Object {
                "fontFamily": "Arial",
              },
            },
            "key": "iohk-dark",
          },
          Object {
            "config": Object {
              "colors": Object {
                "primary": "thistle",
                "secondary": "dodgerblue",
              },
              "typography": Object {
                "fontFamily": "Helvetica",
              },
            },
            "key": "iohk-light",
          },
        ],
      }
    }
    id="consumer-value-holder"
  />
</Provider>
`;

exports[`Theme context <Provider /> it renders the provider component correctly 1`] = `
<ContextProvider
  value={
    Object {
      "key": "iohk-dark",
      "originalTheme": Object {
        "config": Object {
          "colors": Object {
            "primary": "tomato",
            "secondary": "papayawhip",
          },
          "typography": Object {
            "fontFamily": "Arial",
          },
        },
        "key": "iohk-dark",
      },
      "setTheme": [Function],
      "theme": Object {
        "colors": Object {
          "primary": "tomato",
          "secondary": "papayawhip",
        },
        "typography": Object {
          "fontFamily": "Arial",
        },
      },
      "themes": Array [
        Object {
          "config": Object {
            "colors": Object {
              "primary": "tomato",
              "secondary": "papayawhip",
            },
            "typography": Object {
              "fontFamily": "Arial",
            },
          },
          "key": "iohk-dark",
        },
        Object {
          "config": Object {
            "colors": Object {
              "primary": "thistle",
              "secondary": "dodgerblue",
            },
            "typography": Object {
              "fontFamily": "Helvetica",
            },
          },
          "key": "iohk-light",
        },
      ],
    }
  }
>
  <p>
    Theme provider children
  </p>
</ContextProvider>
`;
